### PR TITLE
Adds new function to sync a single Amazon purchase with our backend

### PR DESCRIPTION
**Pull request by vegaro**
Same as https://github.com/RevenueCat/purchases-android/pull/470, but against main

### CODE REVIEW

**Comment by vegaro** [comment]
This will go in 5.1.0

**Comment by vegaro** [comment]
Updated to add currency code and price. This should be ready to go

File: feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt [hunk 1]
```diff
@@ -107,6 +111,39 @@ internal class AmazonBilling constructor(
...
+    override fun normalizePurchaseData(
+        productID: String,
+        purchaseToken: String,
+        storeUserID: String,
+        onSuccess: (correctProductID: String) -> Unit,
+        onError: (PurchasesError) -> Unit
+    ) {
+        val currentlyCachedTokensToSkus = cache.getReceiptSkus()
+
+        currentlyCachedTokensToSkus[purchaseToken]?.let { sku ->
```
Comment (beylmk) [inline feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt:123]: do we have to worry about skus that were cached before this PR not being in the right format or anything?

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 2]
```diff
@@ -268,6 +249,113 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
         )
     }
 
+    private fun syncPurchaseWithBackend(
```
Comment (beylmk) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:252]: nit: private functions at bottom of file

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 3]
```diff
@@ -268,6 +249,113 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
...
+     * The receipt IDs are cached if successfully posted so they are not posted more than once.
+     *
+     * @param [productID] Product ID associated to the purchase.
+     * @param [receiptId] ReceiptId that represents the Amazon purchase.
+     * @param [amazonUserID] Amazon's userID. This parameter will be ignored when syncing a Google purchase.
+     * @param [isoCurrencyCode] Product's currency code in ISO 4217 format.
+     * @param [price] Product's price.
+     */
+    fun syncObserverModeAmazonPurchase(
+        productID: String,
```
Comment (beylmk) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:308]: can we normalize on ID being fully uppercase?

File: feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt (hunk 1 above)
Comment (vegaro) [reply feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt:123]: We are not changing the format, are we?

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 4]
```diff
@@ -268,6 +249,113 @@ class Purchases @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE) intern
...
+                    appUserID,
+                    unsyncedSubscriberAttributesByKey,
+                    body.getAttributeErrors()
+                )
+                deviceCache.addSuccessfullyPostedToken(purchaseToken)
+                cacheCustomerInfo(info)
+                sendUpdatedCustomerInfoToDelegateIfChanged(info)
+                onSuccess()
+            },
+            onError = { error, errorIsFinishable, body ->
```
Comment (beylmk) [inline purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:281]: i think this was here before, but i'm a little confused what "errorIsFinishable" means... it looks like the callback type calls it "shouldConsumePurchase"? 

so is it representing whether a purchase was successful?

Comment (vegaro) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt]: it represents whether the purchase should be consumed. It's bad naming copied from iOS. I can rename to `shouldConsumePurchase`

File: feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt (hunk 1 above)
Comment (beylmk) [reply feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt:123]: didn't think so, was just confirming!

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 4 above)
Comment (beylmk) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt]: thanks :)

File: purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt (hunk 3 above)
Comment (vegaro) [reply purchases/src/main/kotlin/com/revenuecat/purchases/Purchases.kt:265]: oh shoot, `receiptId` should be `receiptID` right? Ugh... good catch!

**Review by beylmk — APPROVED** [review]
🔄

**Pull request merged**